core/src/main/java/brooklyn/util/crypto/SecureKeys.java

```diff
@@ -19,182 +19,149 @@
-            throw Exceptions.propagate(e);
-        }
-        keyPairGenerator.initialize(bits);
-        return keyPairGenerator;
+    public static class PassphraseProblem extends IllegalStateException {
+        private static final long serialVersionUID = -3382824813899223447L;
+        public PassphraseProblem() { super("Passphrase problem with this key"); }
+        public PassphraseProblem(String message) { super("Passphrase problem with this key: "+message); }
     }
     
-    public static KeyPair newKeyPair() {
-        return defaultKeyPairGenerator.generateKeyPair();
+    private SecureKeys() {}
+    
+    /** RFC1773 order, with None for other values. Normally prefer X500Principal. */
+    public static X509Principal getX509PrincipalWithCommonName(String commonName) {
+        return new X509Principal("" + "C=None," + "L=None," + "O=None," + "OU=None," + "CN=" + commonName);
     }
 
-    public static KeyPair newKeyPair(String algorithm, int bits) {
-        return newKeyPairGenerator(algorithm, bits).generateKeyPair();
-    }
+    /** reads RSA or DSA / pem style private key files (viz {@link #toPem(KeyPair)}), extracting also the public key if possible
+     * @throws IllegalStateException on errors, in particular {@link PassphraseProblem} if that is the problem */
+    public static KeyPair readPem(InputStream input, final String passphrase) {
+        // TODO cache is only for fallback "reader" strategy 
+        byte[] cache = Streams.readFully(input);
+        input = new ByteArrayInputStream(cache);
 
-    /** returns a new keystore, of the default type, and initialized to be empty.
-     * (everyone always forgets to load(null,null).) */
-    public static KeyStore newKeyStore() {
         try {
-            KeyStore store = KeyStore.getInstance(KeyStore.getDefaultType());
-            store.load(null, null);
-            return store;
-        } catch (Exception e) {
-            throw Exceptions.propagate(e);
-        }
-    }
+            PEMParser pemParser = new PEMParser(new InputStreamReader(input));
+
+            Object object = pemParser.readObject();
+            pemParser.close();
+
+            JcaPEMKeyConverter converter = new JcaPEMKeyConverter().setProvider("BC");
+            KeyPair kp = null;
+            if (object==null) {
+                throw new IllegalStateException("PEM parsing failed: missing or invalid data");
+            } else if (object instanceof PEMEncryptedKeyPair) {
+                if (passphrase==null) throw new PassphraseProblem("passphrase required");
+                try {
+                    PEMDecryptorProvider decProv = new JcePEMDecryptorProviderBuilder().build(passphrase.toCharArray());
+                    kp = converter.getKeyPair(((PEMEncryptedKeyPair) object).decryptKeyPair(decProv));
+                } catch (Exception e) {
+                    Exceptions.propagateIfFatal(e);
+                    throw new PassphraseProblem("wrong passphrase");
```

Comment:
I'd pass in the cause as well. Is there any more specific exception we can catch here? Feels risky that we could get other exceptions and swallow them, incorrectly reporting it as a "wrong passphrase".